src/catalog/catalog.cpp

```diff
@@ -30,413 +30,764 @@ Catalog *Catalog::GetInstance(void) {
-  auto database_catalog = CreateDatabaseCatalog(START_OID,
-      DATABASE_CATALOG_NAME);
-  storage::DataTable *databases_table = database_catalog.release();
-  database->AddTable(databases_table, true);
-  auto table_catalog = CreateTableCatalog(START_OID, TABLE_CATALOG_NAME);
-  storage::DataTable *tables_table = table_catalog.release();
-  database->AddTable(tables_table, true);
-  databases_.push_back(database);
-  LOG_TRACE("Catalog database created");
-}
+  // TODO: This should be deprecated, dbname should only exists in pg_db
+  database->setDBName(database_name);
 
-// Create a database
-ResultType Catalog::CreateDatabase(std::string database_name,
-    concurrency::Transaction *txn) {
-  // Check if a database with the same name exists
-  for (auto database : databases_) {
-    if (database->GetDBName() == database_name) {
-      LOG_TRACE("Database already exists. Returning ResultType::FAILURE.");
-      return ResultType::FAILURE;
-    }
+  {
+    std::lock_guard<std::mutex> lock(catalog_mutex);
+    databases_.push_back(database);
   }
-  oid_t database_id = GetNextOid();
-  storage::Database *database = new storage::Database(database_id);
-  database->setDBName(database_name);
-  databases_.push_back(database);
 
-  InsertDatabaseIntoCatalogDatabase(database_id, database_name, txn);
+  // Insert database tuple
+  pg_database->InsertDatabase(database_oid, database_name, pool_.get(), txn);
 
   LOG_TRACE("Database created. Returning RESULT_SUCCESS.");
   return ResultType::SUCCESS;
 }
 
-void Catalog::AddDatabase(storage::Database *database) {
-  databases_.push_back(database);
-  std::string database_name;
-  InsertDatabaseIntoCatalogDatabase(database->GetOid(), database_name, nullptr);
-}
-
-void Catalog::AddDatabase(std::string database_name,
-                          storage::Database *database){
-  database->setDBName(database_name);
-  databases_.push_back(database);
-  LOG_DEBUG("Added database with name: %s", database_name.c_str());
-  InsertDatabaseIntoCatalogDatabase(database->GetOid(), database_name, nullptr);
-}
-
+ResultType Catalog::CreateTable(const std::string &database_name,
+                                const std::string &table_name,
+                                std::unique_ptr<catalog::Schema> schema,
+                                concurrency::Transaction *txn) {
+  auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
+  bool single_statement_txn = false;
 
-void Catalog::InsertDatabaseIntoCatalogDatabase(oid_t database_id,
-    std::string &database_name, concurrency::Transaction *txn) {
-  // Update catalog_db with this database info
-  auto tuple =
-      GetDatabaseCatalogTuple(
-          databases_[START_OID]->GetTableWithName(DATABASE_CATALOG_NAME)->GetSchema(),
-          database_id, database_name, pool_);
-  catalog::InsertTuple(
-      databases_[START_OID]->GetTableWithName(DATABASE_CATALOG_NAME),
-      std::move(tuple), txn);
-}
+  if (txn == nullptr) {
```

Comment:
The reasons we start the txn for a calling method are:
(1) some of the test methods just use nullptr for convenience, but we got that fixed.
(2) in initialization part(https://github.com/cmu-db/peloton/blob/master/src/common/init.cpp#L69), it will pass nullptr to create database(since there is no transaction going right now). But I think you're right that we should force anyone to start txn before calling functions in catalog.